dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/requests/McpToolCallParams.java

```diff
@@ -74,22 +76,45 @@ public Object[] getArguments(Jsonb jsonb) {
 
     private Object[] parseArguments(JsonObject arguments2, Jsonb jsonb) {
         JsonObject argsObject = arguments2.asJsonObject();
-        Object[] results = new Object[metadata.arguments().size()];
+        int resultSize = (metadata.arguments() != null ? metadata.arguments().size() : 0) +
+                         (metadata.specialArguments() != null ? metadata.specialArguments().size() : 0);
+        Object[] results = new Object[resultSize];
         HashSet<String> argsProcessed = new HashSet<>();
-        for (var entry : argsObject.entrySet()) {
-            String argName = entry.getKey();
-            JsonValue argValue = entry.getValue();
-            ArgumentMetadata argMetadata = metadata.arguments().get(argName);
-            if (argMetadata != null) {
-                String json = jsonb.toJson(argValue);
-                results[argMetadata.index()] = jsonb.fromJson(json, argMetadata.type());
+        if (metadata.arguments() != null) {
+            for (var entry : argsObject.entrySet()) {
+                String argName = entry.getKey();
+                JsonValue argValue = entry.getValue();
+                ArgumentMetadata argMetadata = metadata.arguments().get(argName);
+                if (argMetadata != null) {
+                    String json = jsonb.toJson(argValue);
+                    results[argMetadata.index()] = jsonb.fromJson(json, argMetadata.type());
+                }
+                argsProcessed.add(argName);
+            }
+
+            if (!argsProcessed.equals(metadata.arguments().keySet())) {
+                List<String> data = generateArgumentMismatchData(argsProcessed, metadata.arguments().keySet());
+                throw new JSONRPCException(JSONRPCErrorCode.INVALID_PARAMS, data);
+            }
+        }
+
+        if (metadata.specialArguments() == null) {
+            return results;
+        }
+
+        HashSet<String> specialArgsProcessed = new HashSet<>();
+        for (var entry : metadata.specialArguments().entrySet()) {
+            SpecialArgumentMetadata specialArgsMetadata = entry.getValue();
 
+            if (specialArgsMetadata.type().equals(Cancellation.class)) {
+                //Cancellation is the only Special Argument type allowed for now
+                results[specialArgsMetadata.index()] = new RequestCancellation();
             }
-            argsProcessed.add(argName);
+            specialArgsProcessed.add(entry.getKey());
         }
 
-        if (!argsProcessed.equals(metadata.arguments().keySet())) {
-            List<String> data = generateArgumentMismatchData(argsProcessed, metadata.arguments().keySet());
+        if (!specialArgsProcessed.equals(metadata.specialArguments().keySet())) {
+            List<String> data = generateArgumentMismatchData(specialArgsProcessed, metadata.specialArguments().keySet());
             throw new JSONRPCException(JSONRPCErrorCode.INVALID_PARAMS, data);
         }
```

Comment:
Let's do this processing elsewhere.

- McpToolCallParams holds the parameters passed by the user, and this processing doesn't use any of that data.
- Doing it this way means we have to put in a placeholder `RequestCancellation` object which we then later replace.

Instead, allow this method to return an array which contains `null` for any parameters which aren't filled in from the request. We can still guarantee that we cover all the arguments as long as:
- We process every regular argument
- We process every special argument
- We check in ToolMetadata.create that every argument is either a regular or special argument